bundle/edu.gemini.pot/src/main/java/edu/gemini/spModel/target/env/AsterismType.java

```diff
@@ -91,24 +93,51 @@ public String displayValue() {
         return displayName;
     }
 
+    public static final SortedSet<AsterismType> DEFAULT_SET;
+
+    static {
+        final SortedSet<AsterismType> s = new TreeSet<>();
+        s.add(Single);
+        DEFAULT_SET = Collections.unmodifiableSortedSet(s);
+    }
+
+    public static final SortedSet<AsterismType> GHOST_SET;
+
+    static {
+        final SortedSet<AsterismType> s = new TreeSet<>();
+        s.add(GhostSingleTarget);
+        s.add(GhostDualTarget);
+        s.add(GhostTargetPlusSky);
+        s.add(GhostSkyPlusTarget);
+        s.add(GhostHighResolutionTargetPlusSky);
+        s.add(GhostHighResolutionTargetPlusSkyPrv);
+        GHOST_SET = Collections.unmodifiableSortedSet(s);
+    }
+
     // Return the asterism types supported by the different instruments.
     // We need to do this here because we want these statically accessible.
     public static SortedSet<AsterismType> supportedTypesForInstrument(final Instrument instType) {
-        final SortedSet<AsterismType> result;
-        if (instType == Instrument.Ghost) {
-            final SortedSet<AsterismType> s = new TreeSet<>();
-            s.add(GhostSingleTarget);
-            s.add(GhostDualTarget);
-            s.add(GhostTargetPlusSky);
-            s.add(GhostSkyPlusTarget);
-            s.add(GhostHighResolutionTargetPlusSky);
-            s.add(GhostHighResolutionTargetPlusSkyPrv);
-            result = Collections.unmodifiableSortedSet(s);
-        } else {
-            final SortedSet<AsterismType> s = new TreeSet<>();
-            s.add(Single);
-            result = Collections.unmodifiableSortedSet(s);
-        }
-        return result;
+        return (instType == Instrument.Ghost) ? GHOST_SET : DEFAULT_SET;
+    }
+
+    public static Option<SortedSet<AsterismType>> supportedTypesForComponent(final SPComponentType compType) {
+        return Instrument
+                .fromComponentType(compType)
+                .map(AsterismType::supportedTypesForInstrument);
+    }
+
+    public static AsterismType forTargetEnvironment(TargetEnvironment env) {
+        return env.getAsterism().asterismType();
+    }
+
+    public static AsterismType forObservation(ISPObservation obs) {
+        return ImOption
+            .apply(SPTreeUtil.findTargetEnvNode(obs))
+            .map(n -> forTargetEnvironment(((TargetObsComp) n.getDataObject()).getTargetEnvironment()))
+            .orElse(ImOption
+                .apply(SPTreeUtil.findInstrument(obs))
+                .filter(n -> n.getDataObject() instanceof SPInstObsComp)
+                .map(n -> ((SPInstObsComp) n.getDataObject()).getPreferredAsterismType())
+            ).getOrElse(AsterismType.Single);
```

Comment:
This is where we figure out the asterism type, checking first the target environment and then falling back to the "preferred" type recorded in the instrument.